src/main/python/run_trecqa.py

```diff
@@ -5,219 +5,294 @@
+                label = label.group(1)
+                label = 1 if label == 'positive' else 0
+                answer = line.lower().split('\t')
+                answer_count += 1
 
-        answer_id = 'Q{}-A{}'.format(qid, answer_count)
-        answers[qid].append((answer_id, answer, label))
-        labels[qid][answer_id] = label
-      prev = line
+                answer_id = 'Q{}-A{}'.format(qid, answer_count)
+                answers[qid].append((answer_id, answer, label))
+                labels[qid][answer_id] = label
+            prev = line
 
-  return questions, answers, labels
+    return questions, answers, labels
 
 
-def create_ground_truth(fname, answers):
-  with open(fname, 'w') as f:
-    for qid in answers.keys():
-      for answer in answers[qid]:
-        f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
+def create_qrel_jaccard(fname, answers):
+    with open(fname, 'w') as f:
+        for qid in answers.keys():
+            for answer in answers[qid]:
+                f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
 
+# evaluate by TREC QA patterns
+def eval_by_pattern(qid, candidates, pattern, out):
+    seen = set([])
 
-def get_precision(actual, predicted, k):
-  total_count = 0.0
-  hits = 0.0
+    for i, cand in enumerate(candidates):
+        this_candidate = " ".join(cand[0])
 
-  for qid in predicted.keys():
-    run_labels = predicted.copy()
+        hash_value = hashlib.md5(this_candidate.encode()).hexdigest()
+        answer_id = "QA{}.{}".format(qid, hash_value)
 
-    if len(predicted[qid]) > k:
-      run_labels = run_labels[qid][:k]
-    else:
-      run_labels = run_labels[qid]
+        if answer_id in seen:
+            continue
+        seen.add(answer_id)
 
-    for p in run_labels:
-      total_count += 1
+        out.write('{} Q0 {} {} {} TrecQA-pattern\n'.format(qid, answer_id, i+1, 1.0/(i + 1)))
 
-      if not actual[qid]:
-        continue
+# evaluate the run file at different depths
+def evaluate_at_k(fname, eval_depth):
+    for k in eval_depth:
+        with open(fname) as f,  open("{}.k{}.txt".format(fname, k), 'w') as output_file:
+            for line in f:
+                det = line.strip().split()
+                rank = int(det[3])
 
-      if "unjudged" in p:
-        hits += 0
-      else:
-        hits += actual[qid][p]
+                if rank <= int(k):
+                    output_file.write(line)
 
-  if total_count > 0:
-    return hits / total_count
-  else:
-    return 0
 
+def create_qrel_pattern(all_sentences, pattern, qrels, qid):
+    seen = set([])
+    for i, cand in enumerate(all_sentences):
+        this_candidate = cand.lower().split("\t")[0]
+        result = re.findall(r'{}'.format(pattern.lower()), this_candidate)
+        hash_value = hashlib.md5(this_candidate.encode()).hexdigest()
+        answer_id = "QA{}.{}".format(qid, hash_value)
+        if answer_id in seen:
+            continue
+        seen.add(answer_id)
+
+        if result:
```

Comment:
You can collapse these, using `1 if result else 0` inside the format ... I believe it's more pythonic, it certainly reduces the possibility of changing one format, but not the other. I'd also suggest using `len(result)` just to be a bit more explicit that result is not a bool